### DISCUSSION TITLE
Make pending jobs scheduled

### DISCUSSION THREAD

**Discussion by uniquefine** [riverqueue/river#393] [Q&A] [answered]
There is an option to insert a [job as pending](https://github.com/riverqueue/river/blob/559259949dbdfeca69fc3819a4fded121cedcdc9/insert_opts.go#L26-L31).
How can you move a job from pending to scheduled once you want it to be worked on?

As additional information for my use case:
I have jobs of different types inserted by multiple concurrent clients.
For each type I want to have one job worked on at a time.
My idea is to insert jobs as pending and to create a periodic job that checks for a given type whether a job is already running and if not transitions a pending job to scheduled.

**Comment by bgentry** [comment]
You would need to create your own logic to make these updates. As it stands `pending` is essentially just a placeholder state for jobs which will never run until they are moved to `available` or `scheduled` by some external process.

Can you give some more info on the use case? It sounds like you want to ensure that only one job of a given job kind is allowed to be active at any given moment. Is that correct?  We’ve been talking about some stuff to potentially solve this.

Are there any other dimensions on which you want to apply concurrency limits like this, such as specific job arguments? Or is it strictly about the job type?

**Reply by uniquefine** [reply]
Thanks for the quick answer, that makes sense. 

> It sounds like you want to ensure that only one job of a given job kind is allowed to be active at any given moment.
Yes, that's exactly it.

For the usecase, the word "type" was very imprecise. It's not about the "kind" of job (as returned by the `Kind()` function of the job args.
We have jobs that do operations on remote servers (identified by an URL). We only want one operation per server to be in progress. But we want operations that are not on the same server to run in parallel. E.g. like this

Server 1: update(_pending_) update(_pending_) config(_in progress_)
Server 2: config(_pending_) update(_in progress_)
Server 3: update(_pending_) config(_pending_) update(_in progress_)

If more jobs are added for the same server they should be queued and started when the previous job finished (or failed).
There are kinds of jobs for different operations e.g. updating a software package on the server or placing a config file there.

I guess you could model this with one queue per server.

**Reply by uniquefine** [reply]
Actually I found the name for this usecase: [Sequential Convoy Pattern](https://learn.microsoft.com/en-us/azure/architecture/patterns/sequential-convoy)

**Reply by brandur** [reply, marked as answer]
FWIW, this is very close to #380. I think we should probably support this, but it's not supported right now. See that issue for a suggested workaround.

**Reply by brandur** [reply]
(And interesting to know an official name for this pattern BTW.)

**Reply by uniquefine** [reply]
Thanks, I didn't see this issue before (but arrived at a similar workaround).
